### PR TITLE
[AMDGPU] Fix folding clamp into pseudo scalar instructions

### PR DESCRIPTION
**Pull request by mbrkusanin**
Clamp is canonically a v_max* instruction with a VGPR dst. Folding clamp
into a pseudo scalar instruction can cause issues due to a change in
regbank. We fix this with a copy.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/si-fold-scalar-clamp.mir [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    ; CHECK-LABEL: name: test
+    ; CHECK: liveins: $sgpr0
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:sgpr_32 = COPY $sgpr0
+    ; CHECK-NEXT: [[V_S_RSQ_F32_e64_:%[0-9]+]]:sgpr_32 = nofpexcept V_S_RSQ_F32_e64 0, [[COPY]], 1, 0, implicit $mode, implicit $exec
+    ; CHECK-NEXT: [[COPY1:%[0-9]+]]:vgpr_32 = COPY [[V_S_RSQ_F32_e64_]]
+    ; CHECK-NEXT: EXP_DONE 0, killed [[COPY1]], [[COPY1]], [[COPY1]], [[COPY1]], -1, 0, 15, implicit $exec
+    ; CHECK-NEXT: S_ENDPGM 0
+    %0:sgpr_32 = COPY $sgpr0
+    %1:sgpr_32 = nofpexcept V_S_RSQ_F32_e64 0, %0, 0, 0, implicit $mode, implicit $exec
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-fold-scalar-clamp.mir:19]: Why does this have a V_ prefix if it's just an S_ instruction?

File: llvm/test/CodeGen/AMDGPU/si-fold-scalar-clamp.mir [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    %2:vgpr_32 = nofpexcept V_MAX_F32_e64 0, %1, 0, %1, -1, 0, implicit $mode, implicit $exec
+    EXP_DONE 0, killed %2, %2, %2, %2, -1, 0, 15, implicit $exec
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-fold-scalar-clamp.mir:21]: Could use a simpler use, just copy to $vgpr0?

**Review by arsenm — APPROVED** [review]
LGTM with optional test nit

File: llvm/test/CodeGen/AMDGPU/si-fold-scalar-clamp.mir (hunk 2 above)
Comment (mbrkusanin) [reply llvm/test/CodeGen/AMDGPU/si-fold-scalar-clamp.mir:21]: Yes for representing the change in patch, but that use case was not previously crashing with -verify-machineinstrs. I wanted something that must use a vgpr. Should I change it anyway?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/si-fold-scalar-clamp.mir:21]: any VALU instruction will do

**Review by arsenm — APPROVED** [review]

**Review by jayfoad — APPROVED** [review]
Thanks!

**Comment by jayfoad** [comment]
@mbrkusanin please get this backported to 19.x.

**Pull request merged**